client/x/evmstaking/keeper/deposit.go

```diff
@@ -96,8 +99,78 @@ func (k Keeper) ProcessDeposit(ctx context.Context, ev *bindings.IPTokenStakingD
 	}
 
 	val, err := k.stakingKeeper.GetValidator(cachedCtx, validatorAddr)
+
+	//nolint:nestif // nested ifs error handling
 	if errors.Is(err, stypes.ErrNoValidatorFound) {
-		return errors.WrapErrWithCode(errors.ValidatorNotFound, err)
+		log.Debug(cachedCtx, "Validator not found, refunding deposit minus the refund fee",
+			"val_story", validatorAddr,
+			"val_pubkey", validatorPubkey.String(),
+		)
+
+		// the min refund fee amount will be `refundFeeBps * minDelegationAmount (1024) / 10_000bps`
+		refundFeeBps, err := k.RefundFeeBps(cachedCtx)
+		if err != nil {
+			return errors.Wrap(err, "get refund fee")
+		}
+		refundFeeAmount := amountCoin.Amount.Mul(math.NewInt(int64(refundFeeBps))).Quo(math.NewInt(10_000))
+		refundAmount := amountCoin.Amount.Sub(refundFeeAmount)
+
+		// refund completion time is 1 day after the deposit
+		completionTime := sdkCtx.BlockTime().Add(time.Hour * 24)
+
+		defer func() {
+			if r := recover(); r != nil {
+				err = errors.WrapErrWithCode(errors.UnexpectedCondition, fmt.Errorf("panic caused by %v", r))
+			}
+
+			var e sdk.Event
+			if err == nil {
+				writeCache()
+				e = sdk.NewEvent(
+					types.EventTypeDelegationRefundSuccess,
+				)
+			} else {
+				e = sdk.NewEvent(
+					types.EventTypeDelegationRefundFailure,
+					sdk.NewAttribute(types.AttributeKeyErrorCode, errors.UnwrapErrCode(err).String()),
+				)
+			}
+
+			sdkCtx.EventManager().EmitEvents(sdk.Events{
+				e.AppendAttributes(
+					sdk.NewAttribute(types.AttributeKeyAmount, ev.StakeAmount.String()),
+					sdk.NewAttribute(types.AttributeKeyRefundAmount, refundAmount.String()),
+					sdk.NewAttribute(types.AttributeKeyBlockHeight, strconv.FormatInt(sdkCtx.BlockHeight(), 10)),
+					sdk.NewAttribute(types.AttributeKeyDelegatorAddress, ev.Delegator.String()),
+					sdk.NewAttribute(types.AttributeKeyValidatorCmpPubKey, hex.EncodeToString(ev.ValidatorCmpPubkey)),
+					sdk.NewAttribute(types.AttributeKeySenderAddress, ev.OperatorAddress.Hex()),
+					sdk.NewAttribute(types.AttributeKeyTxHash, hex.EncodeToString(ev.Raw.TxHash.Bytes())),
+					sdk.NewAttribute(types.AttributeKeyRefundCompletionTime, completionTime.String()),
+				),
+			})
+		}()
+
+		// push the refund to the unbonding queue
+		_, err = k.stakingKeeper.SetUnbondingDelegationEntry(
```

Comment:
It is not actual unbonding, so I think we need to use `stakingKeeper.InsertUBDQueue` to add it to the unbonding queue and it's enough ([ref](https://github.com/piplabs/cosmos-sdk/blob/40741b13972bbdfd51568a37e5a7f7d397985782/x/staking/keeper/delegation.go#L492)). `SetUnbondingDelegationEntry` is a function that stores the unbonding delegation using an index that increments by 1, without adding it to the unbonding queue.

Reply:
fixed in f027788

If I understand your request correctly, we need to add `InsertUBDQueue` (add to queue) after `SetUnbondingDelegationEntry` (create a ubd with unique ID).

This is following `Undelegate` fn which calls `SetUnbondingDelegationEntry` and then `InsertUBDQueue` ([ref](https://github.com/piplabs/cosmos-sdk/blob/40741b13972bbdfd51568a37e5a7f7d397985782/x/staking/keeper/delegation.go#L1270-L1280)).

Reply:
What if we do not call `SetUnbondingDelegationEntry`? IIUC, it should be fine. So I thought that we can just call `InsertUBDQueue`.

Reply:
Ah, I thought the unbonding queue was independent of unbonding delegation. But as you mentioned, it seems to be required since our EndBlocker process withdraws by dequeuing and completing the unbonding. SGTM. We should call both 👍